sonar-db/src/main/java/org/sonar/db/charset/MssqlCharsetHandler.java

```diff
@@ -0,0 +1,139 @@
+    // Example of row:
+    // issues | kee | Latin1_General_CS_AS
+    List<ColumnDef> columns = select(connection,
+      ColumnDef.SELECT_COLUMNS +
+        "FROM [INFORMATION_SCHEMA].[COLUMNS] " +
+        "WHERE collation_name is not null " +
+        "ORDER BY table_name,column_name", ColumnDef.ColumnDefRowConverter.INSTANCE);
+    for (ColumnDef column : columns) {
+      if (!endsWithIgnoreCase(column.getCollation(), "_CS_AS")) {
+        repairColumnCollation(connection, column);
+      }
+    }
+  }
+
+  private void repairColumnCollation(Connection connection, ColumnDef column) throws SQLException {
+    // 1. select the indices defined on this column
+    String selectIndicesSql = format("SELECT I.name as index_name, I.is_unique as unik, IndexedColumns " +
+      "     FROM sys.indexes I " +
+      "     JOIN sys.tables T ON T.Object_id = I.Object_id " +
+      "     JOIN (SELECT * FROM ( " +
+      "     SELECT IC2.object_id, IC2.index_id, " +
+      "     STUFF((SELECT ' ,' + C.name " +
+      "     FROM sys.index_columns IC1 " +
+      "     JOIN sys.columns C " +
+      "     ON C.object_id = IC1.object_id " +
+      "     AND C.column_id = IC1.column_id " +
+      "     AND IC1.is_included_column = 0 " +
+      "     WHERE IC1.object_id = IC2.object_id " +
+      "     AND IC1.index_id = IC2.index_id " +
+      "     GROUP BY IC1.object_id,C.name,index_id " +
+      "     ORDER BY MAX(IC1.key_ordinal) " +
+      "     FOR XML PATH('')), 1, 2, '') IndexedColumns " +
+      "     FROM sys.index_columns IC2 " +
+      "     GROUP BY IC2.object_id ,IC2.index_id) tmp1 )tmp2 " +
+      "     ON I.object_id = tmp2.object_id AND I.Index_id = tmp2.index_id " +
+      "     WHERE I.is_primary_key = 0 AND I.is_unique_constraint = 0 " +
+      "     and T.name =('%s') " +
+      "     and CHARINDEX ('%s',IndexedColumns)>0", column.getTable(), column.getColumn());
+    List<ColumnIndex> indices = getSqlExecutor().executeSelect(connection, selectIndicesSql, ColumnIndexConverter.INSTANCE);
+
+    // 2. drop indices
+    for (ColumnIndex index : indices) {
+      getSqlExecutor().executeUpdate(connection, format("DROP INDEX %s.%s", column.getTable(), index.name));
+    }
+
+    // 3. alter collation of column
+    String csCollation = toCaseSensitive(column.getCollation());
+
+    String nullability = column.isNullable() ? "NULL" : "NOT NULL";
+    String size = column.getSize() >= 0 ? String.valueOf(column.getSize()) : "max";
+    String alterSql = format("ALTER TABLE %s ALTER COLUMN %s %s(%s) COLLATE %s %s",
+      column.getTable(), column.getColumn(), column.getDataType(), size, csCollation, nullability);
+    LOGGER.info("Changing collation of column [{}.{}] from {} to {} | sql=", column.getTable(), column.getColumn(), column.getCollation(), csCollation, alterSql);
+    getSqlExecutor().executeUpdate(connection, alterSql);
+
+    // 4. re-create indices
+    for (ColumnIndex index : indices) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Combine this loop with the one that starts on line 90. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3047)